tls/s2n_connection.c

```diff
@@ -1289,6 +1289,10 @@ struct s2n_cert_chain_and_key *s2n_connection_get_selected_cert(struct s2n_conne
 
 uint8_t s2n_connection_get_protocol_version(struct s2n_connection *conn)
 {
+    if (conn == NULL) {
+        return S2N_UNKNOWN_PROTOCOL_VERSION;
+    }
+
```

Comment:
While you're in here.  There are only two possibilities for conn->mode: server and client. https://github.com/awslabs/s2n/blob/7db49613f830e975cff146aab5b4bb862682e772/api/s2n.h#L173
The code should error in this case, not keep going.

Reply:
Shouldn't this function return an error signal instead of `S2N_UNKNOWN_PROTOCOL_VERSION`?

Doing a quick search reveals that both `>` and `<` comparisons are being used on the return value of this. `S2N_UNKNOWN_PROTOCOL_VERSION` being defined as `0` means it will be `true` for the `<` checks.

https://github.com/awslabs/s2n/blob/a38833075264f493e4a1fa126bb88cfc3fb50142/tls/extensions/s2n_client_signature_algorithms.c#L40
https://github.com/awslabs/s2n/blob/ddf3cd06558fe2126e8436c539422789513e5777/tls/extensions/s2n_server_session_ticket.c#L38

Reply:
This function is intended as a light-weight replacement for developers needing to figure out whether they should be using actual, client, or server _protocol_version. We do a lot of conn->client/server/actual_protocol_version comparisons all over our code, and we never check for S2N_UNKNOWN_PROTOCOL_VERSION. Basically, existing logic treats S2N_UNKNOWN_PROTOCOL_VERSION as the oldest possible version of tls.